### PR TITLE
UI Graphical Button

### PR DESCRIPTION
**Pull request by nhaagen**
The graphical button combines the recognisability of a graphical element with the explicity of a text-label. It has an obtrusive presentation that indicate import actions on the screen and make them hard to overlook.

This PR also introduces some changes to the rendering of glyphs, depending on context: a glyph does not necessarily need to be rendered as an < a >-tag.

Rules: https://github.com/ILIAS-eLearning/ILIAS/pull/882/files#diff-b7791ca027a51c7e6e50c3abefeb4394

graphicalbutton

### CODE REVIEW

**Comment by klees** [comment]
I will do some adjustments to the text to incorporate the "Trigger Area" from the Page Editor Revision.

**Comment by klees** [comment]
Hi @alex40724, hi @Amstutz,
I incorporated one sentence from the trigger area and added some description and rules. Please give feedback.
I also have another proposal for a name: simply "Huge Button".
Regards!

File: Services/Init/classes/class.ilInitialisation.php [hunk 1]
```diff
@@ -1359,6 +1359,12 @@ protected static function initUIFramework(\ILIAS\DI\Container $c) {
 					( $c["ui.resource_registry"]
 					, new ILIAS\UI\Implementation\Render\FSLoader
 						( new ILIAS\UI\Implementation\Render\DefaultRendererFactory
+							($c["ui.factory"]
+							, $c["ui.template_factory"]
+							, $c["lng"]
+							, $c["ui.javascript_binding"]
+							),
+						  new ILIAS\UI\Implementation\Component\Glyph\GlyphRendererFactory
```
Comment (Amstutz) [inline Services/Init/classes/class.ilInitialisation.php:1366]: Why is this necessary?

File: src/UI/Component/Button/Factory.php [hunk 2]
```diff
@@ -209,5 +209,55 @@ public function month($default);
 	 */
 	public function tag($label, $action);
 
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *     The graphical button combines the recognisability of a graphical element
+	 *     with the explicity of a text-label. It has an obtrusive presentation that
+	 *     indicate important actions on the screen and make them hard to overlook.
```
Comment (Amstutz) [inline src/UI/Component/Button/Factory.php]: +1

File: src/UI/Component/Button/Factory.php [hunk 3]
```diff
@@ -209,5 +209,55 @@ public function month($default);
...
+	 *   usage:
+	 *     1: >
+	 *       Since Graphical Buttons are so obtrusive they MUST only be used
+	 *       to indicate important actions on the screen.
+	 *   wording:
+	 *     1: The icon/glyph and the text on the Graphical Button MUST be corresponding.
+	 *   style:
+	 *     1: >
+	 *       Graphical Buttons MUST occupy as much space as their container
+	 *       leaves them.
```
Comment (Amstutz) [inline src/UI/Component/Button/Factory.php]: +1, this is good and important. But currently not yet the case, right. At least in the examples there is no widht: 100% or similar to be found in the according CSS.

File: src/UI/Component/Button/Factory.php [hunk 4]
```diff
@@ -209,5 +209,55 @@ public function month($default);
...
+	 *     2: >
+	 *       When used to toggle the visibility of another element,
+	 *       the button MUST reflect the elements state of visibility.
```
Comment (Amstutz) [inline src/UI/Component/Button/Factory.php]: We have that with other buttons as well, right? Maybe even promote that to the Button Class.

File: src/UI/Component/Button/Graphical.php [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * This describes an graphical button.
+ */
+interface Graphical extends Button {
+
+	/**
+	 * @return ILIAS\UI\Component\Icon\Icon | \ILIAS\UI\Component\Glyph\Glyph
+	 */
+	public function getIconOrGlyph();
```
Comment (Amstutz) [inline src/UI/Component/Button/Graphical.php]: Please add description.

File: src/UI/Component/Button/Graphical.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+	/**
+	 * @param 	bool 	$state
+	 * @return 	Graphical
+	 */
+	public function withEngagedState($state);
```
Comment (Amstutz) [inline src/UI/Component/Button/Graphical.php]: Please add description.

File: src/UI/Component/Button/Graphical.php [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+	/**
+	 * @return 	bool
+	 */
+	public function isEngaged();
```
Comment (Amstutz) [inline src/UI/Component/Button/Graphical.php]: Please add description.

File: src/UI/templates/default/Button/button.less [hunk 8]
```diff
@@ -36,6 +36,17 @@ input.btn{
...
+    &:hover,
+    &:focus,
+    &:active,
+    &.active {
+      background-color: @il-disabled-btn-bg;
+      border-color: @il-disabled-btn-border;
+      color: @il-disabled-btn-color;
+    }
+  }
 }
```
Comment (Amstutz) [inline src/UI/templates/default/Button/button.less:50]: And probably some witdh:100% for btn-graphical.

**Review by Amstutz — CHANGES_REQUESTED** [review]
The interface LGTM in general. Some minor suggestion for improvements.

I do think that this is the same thing as is needed for the page editor button. Nice to see first synergy at work here.

Note there are the missing tests for a final merge after the JF decision.

File: Services/Init/classes/class.ilInitialisation.php (hunk 1 above)
Comment (klees) [reply Services/Init/classes/class.ilInitialisation.php:1366]: Because Glyphs need to be rendered without an `<a>`-Tag when they are contained in a Graphical Button.

File: src/UI/Component/Button/Factory.php (hunk 3 above)
Comment (klees) [reply src/UI/Component/Button/Factory.php]: Yes.

File: src/UI/Component/Button/Factory.php (hunk 4 above)
Comment (klees) [reply src/UI/Component/Button/Factory.php]: Yes, I think so. Is this the same what we had in the "disable [today]"-discussion? (Remember?)

File: src/UI/Component/Button/Graphical.php (hunk 5 above)
Comment (klees) [reply src/UI/Component/Button/Graphical.php]: @nhaagen Would you?

File: src/UI/Component/Button/Graphical.php (hunk 6 above)
Comment (klees) [reply src/UI/Component/Button/Graphical.php]: @nhaagen Would you?

File: src/UI/Component/Button/Graphical.php (hunk 7 above)
Comment (klees) [reply src/UI/Component/Button/Graphical.php]: @nhaagen Would you?

File: src/UI/templates/default/Button/button.less (hunk 8 above)
Comment (klees) [reply src/UI/templates/default/Button/button.less:50]: @nhaagen Would you check if that works?

File: src/UI/Component/Button/Factory.php (hunk 4 above)
Comment (Amstutz) [reply src/UI/Component/Button/Factory.php]: Yes I do remember and feel ashamed, this is still lingering in my todo box.

**Review by alex40724 — APPROVED** [review]
Looks all good to me. I do not see any reasons why this could not be used in the page editor as well. Mostly page editor css would make changes to the design I guess.

**Comment by klees** [comment]
This might be the shortest KS discussion ever.

**Comment by klees** [comment]
Any opinions on "Huge Button"?

**Comment by Amstutz** [comment]
> Any opinions on "Huge Button"?

Yes, I don't like it too much, since this button is not necessarily all that huge to me in all cases.

**Comment by klees** [comment]
Yeah, that's right. But I don't like graphical (and iconographic as well) not that much anymore as well, because the main thing to me is not the icon anymore but that "takes space" thing instead.

**Comment by Amstutz** [comment]
Maybe "Spaced Button" or "Area Button"?

**Comment by klees** [comment]
"Bulky Button"

**Comment by Amstutz** [comment]
+1 for bulky button. I like.

File: src/UI/Component/Button/Factory.php [hunk 9]
```diff
@@ -209,5 +209,55 @@ public function month($default);
 	 */
 	public function tag($label, $action);
 
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *     The graphical button combines the recognisability of a graphical element
```
Comment (atoedt) [inline src/UI/Component/Button/Factory.php]: The graphical button is highly obtrusive combing a glyph and a textual label on an unusually shaped or sized button. 
The graphical button opens a rich and relevant dropdown. (Page Editor case)
Your case: The graphical button toggles "element"  visible / invisible. (as stated below)

File: src/UI/Component/Button/Factory.php [hunk 10]
```diff
@@ -209,5 +209,55 @@ public function month($default);
...
+	 *     with the explicity of a text-label. It has an obtrusive presentation that
+	 *     indicate important actions on the screen and make them hard to overlook.
+	 *
+	 *   composition: >
+	 *     The graphical button consists of an icon or glyph and a (very short) text.
```
Comment (atoedt) [inline src/UI/Component/Button/Factory.php]: pls remove "an icon" 
an icon is colorful

File: src/UI/Component/Button/Factory.php [hunk 11]
```diff
@@ -209,5 +209,55 @@ public function month($default);
...
+	 *
+	 *   effect: >
+	 *     Both, icon/glyph and text, form an entity an can only be operated together.
```
Comment (atoedt) [inline src/UI/Component/Button/Factory.php]: pls delete "Both, icon/glyph and text, form an entity an can only be operated together."

A button is always operated as a whole.

File: src/UI/Component/Button/Factory.php [hunk 12]
```diff
@@ -209,5 +209,55 @@ public function month($default);
...
+	 *     The button has a "engaged"-state: When the button is used to toggle the
```
Comment (atoedt) [inline src/UI/Component/Button/Factory.php]: For toggling please use the accepted View Control Mode. 

Kitchensink clearly defines Buttons as controls that "trigger interactions that change the system’s status."
Toggling does not trigger interaction. If this is about toggling please use the respective accepted control.

File: src/UI/Component/Button/Factory.php [hunk 13]
```diff
@@ -209,5 +209,55 @@ public function month($default);
...
+	 *     visibility of an element, it stays engaged until the element is hidden again.
```
Comment (atoedt) [inline src/UI/Component/Button/Factory.php]: What is meant by "element"? Without enough specificity this cannot be critiqued. Give an example at least

File: src/UI/Component/Button/Factory.php [hunk 14]
```diff
@@ -209,5 +209,55 @@ public function month($default);
...
+	 *   rival:
+	 *     Primary Button: >
+	 *       Primary Buttons indicate the most important action among a collection
+	 *       of actions, e.g. in a tool bar, controls of a form or in a modal. Graphical
+	 *       buttons make it hard to miss the indicated action by occupying space.
+	 *
+	 * rules:
+	 *   usage:
+	 *     1: >
+	 *       Since Graphical Buttons are so obtrusive they MUST only be used
```
Comment (atoedt) [inline src/UI/Component/Button/Factory.php]: This rule does not set graphical buttons aside from primary buttons. This needs sharpening.
Why could one not use graphical buttons in forms or toolbars?

File: src/UI/Component/Button/Factory.php [hunk 15]
```diff
@@ -209,5 +209,55 @@ public function month($default);
...
+	 *       to indicate important actions on the screen.
+	 *   wording:
```
Comment (atoedt) [inline src/UI/Component/Button/Factory.php]: Why are there no rules on interaction? Is everything allowed? 
if you think about toggling you should elaborate this behavior here.

File: src/UI/Component/Button/Factory.php [hunk 16]
```diff
@@ -209,5 +209,55 @@ public function month($default);
...
+	 *     1: The icon/glyph and the text on the Graphical Button MUST be corresponding.
+	 *   style:
+	 *     1: >
+	 *       Graphical Buttons MUST occupy as much space as their container
```
Comment (atoedt) [inline src/UI/Component/Button/Factory.php]: if this has a container what could this be? This should be mentioned in the composition section: What can contain this type of button? Forms and Toolbars?

File: src/UI/Component/Button/Factory.php [hunk 17]
```diff
@@ -209,5 +209,55 @@ public function month($default);
...
+	 *       leaves them.
+	 *     2: >
+	 *       When used to toggle the visibility of another element,
```
Comment (atoedt) [inline src/UI/Component/Button/Factory.php]: What is an element?

File: src/UI/Component/Button/Factory.php (hunk 9 above)
Comment (klees) [reply src/UI/Component/Button/Factory.php]: I incorporated some of your wording suggestions, but not everything.

Opening a dropdown to me actually is toggling visibilty (at least toggling it from of to on) so it is not that different from opening a slate.

File: src/UI/Component/Button/Factory.php (hunk 10 above)
Comment (klees) [reply src/UI/Component/Button/Factory.php]: Icons may indeed be colorful. We need that possibility to be able to put things like "The Repository" or "The Personal Desktop" into the sidebar of the new desktop.

File: src/UI/Component/Button/Factory.php (hunk 11 above)
Comment (klees) [reply src/UI/Component/Button/Factory.php]: Removed that.

File: src/UI/Component/Button/Factory.php (hunk 13 above)
Comment (klees) [reply src/UI/Component/Button/Factory.php]: Changed to "Component". This is a well defined word.

File: src/UI/Component/Button/Factory.php (hunk 14 above)
Comment (klees) [reply src/UI/Component/Button/Factory.php]: "The primary button indicates **the** most important action on a screen.", so there at least is a distinction.

Any idea for the sharpening? I mean basically we need a rule that says if it is ok to make an action obtrusive, right? It needs to cover the Page Layout usecase for this button in the sidebar and the Page Editor usecase as a placeholder for a page element to be added, correct?

File: src/UI/Component/Button/Factory.php (hunk 15 above)
Comment (klees) [reply src/UI/Component/Button/Factory.php]: I don't think this should be different from a button in general. I see that the current button rules do not cover this usecase, but I think they should.

File: src/UI/Component/Button/Factory.php (hunk 16 above)
Comment (klees) [reply src/UI/Component/Button/Factory.php]: The to be created Side Bar and the to be created Component for the Page Layout.

File: src/UI/Component/Button/Factory.php (hunk 12 above)
Comment (klees) [reply src/UI/Component/Button/Factory.php]: I don't think that's correct. Mode View Control switches between different aspects of some data. We don't have that aspect in the Side Bar usecase, as well as we don't have that in the Page Editor usecase.

I think this is a problem of the button definition. After all, the mode view control already uses button for that exact switching purpose, which is a violation of the rules as they currently are. The Dropdown on the other hand does not use the button as a trigger, but it arguably should do that, in general and for the Page Editor use case.

For the Mode View Control and the Section View Control we also already found the need for an engaged state of the button. So imo it becomes clearer that the button indeed can do other stuff than "trigger interactions that change the system’s status" but also may trigger actions that change view's status.

Comment (klees) [reply src/UI/Component/Button/Factory.php]: Summoning @Amstutz.

Comment (klees) [reply src/UI/Component/Button/Factory.php]: Opened https://github.com/ILIAS-eLearning/ILIAS/pull/897 for change in button rules.

**Comment by Amstutz** [comment]
What is the state of this? Is there something missing? Is this JF ready?

**Comment by alex40724** [comment]
@Amstutz  Current state seems to be that you are "requesting changes" :-) Do you think these changes need to be done before JF discussion? I think this is JF ready.

**Comment by Amstutz** [comment]
@alex40724 You are right, and sorry, I was not clear enough. There are some unresolved minor issues on which i commented (about comments of interface method). This is not a deal breaker for me. In any case, this is so close to be JF ready, that I think @klees can add the JF label again, if he feels ready to present it.

**Comment by matthiaskunkel** [comment]
Jour Fixe, 12 MAR 2018: We highly appreciate this PR. Please change the name to "Bulky Button" to address it's obtrusive character.

File: tests/UI/Component/Button/BulkyButtonTest.php [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+			$this->assertTrue(true);
+		}
+		try {
+			$f->bulky("", "label", "http://www.ilias.de");
+			$this->assertFalse("This should not happen");
+		}
+		catch (\InvalidArgumentException $e) {
+			$this->assertTrue(true);
+		}
+	}
```
Comment (klees) [inline tests/UI/Component/Button/BulkyButtonTest.php:61]: Please split this test in three parts.

File: tests/UI/Component/Button/BulkyButtonTest.php [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+		$this->assertEquals(
+			$this->glyph,
+			$b->getIconOrGlyph()
+		);
+		$b = $this->factory->button()->bulky($this->icon, "label", "http://www.ilias.de");
+		$this->assertEquals(
+			$this->icon,
+			$b->getIconOrGlyph()
+		);
+	}
```
Comment (klees) [inline tests/UI/Component/Button/BulkyButtonTest.php:77]: Please split this test in two parts.

File: tests/UI/Component/Button/BulkyButtonTest.php [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+			.'		<span class="glyphicon glyphicon-briefcase" aria-hidden="true"></span>'
+			.'	</span>'
+			.'	<span class="bulky-label">label</span>'
+			.'</button>';
+
+		$this->assertHTMLEquals(
+			$expected,
+			$r->render($b)
+		);
+	}
```
Comment (klees) [inline tests/UI/Component/Button/BulkyButtonTest.php:121]: These are two tests as well.

File: src/UI/templates/default/Button/button.less (hunk 8 above)
Comment (klees) [reply src/UI/templates/default/Button/button.less:50]: @nhaagen Did you try this?

**Review by klees — CHANGES_REQUESTED** [review]
Still LGTM in general only found minor problems in tests. 

@Amstutz: Anything else from your side?

**Comment by Amstutz** [comment]
Still waiting on some resolve to the width 100% comment of mine. Other thant that, I would be fine with a merge.

**Comment by klees** [comment]
Just merged this manually.
